### ISSUE TITLE
Jest snapshot testing

### ISSUE THREAD

**Issue by pmelab** [AmazeeLabs/gatsby-starter-amazee#153]
We should add jest snapshot testing for storybook components. In our latest project, there is a nicely working setup.

**Reasons:**
* In combination with tailwind, snapshot tests become (almost) visual regression tests
* Code coverage on snapshots gives an indication of how many components (and branched variations) are properly reflected in storybook

**Comment by JohnAlbin** [comment]
> In combination with tailwind, snapshot tests become (almost) visual regression tests

This is false. My team tried to use Jest snapshots as a pseudo-visual regression tests on our first decoupled site. We ran into 2 serious problems, only 1 of which can be fixed.

1. About once every week or two, some non-CSS change in the code base would cause all the auto-generated class names to change even though the CSS didn't change. Because of these frequent false positives, developers got used to just re-generating the snapshots without looking too closely. If there were any real unintended CSS changes, these would have been treated the same as false positives. Tailwind doesn't have auto-generated classes, so this point is moot. In PostCSS, you can work around this problem by configuring webpack to use (during snapshot generation) path-based class names instead of the default hash-based class names that were affected by this problem.
2. Since Jest snapshots only capture the class names, the snapshots don't change even when the class' CSS _properties_ change. Specifically, any change you make in `tailwind.config.ts` will not change the results of a snapshot. _This is super bad._ If I change something in the config file mistakenly thinking it will have a small change and it turns out it changes lots of things, the snapshots will never help me discover my mistake. That's exactly what visual regression tests are supposed to do: alert us when we do something stupid.

Since Jest snapshots didn't work as visual regression tests, it meant all the time we spent maintaining and updating them were a literal waste of our time. We ended up removing them from the project.

> Code coverage on snapshots gives an indication of how many components (and branched variations) are properly reflected in storybook

I think I just undermined the assumption this logic is based on. Code coverage is defined as the percentage of code that is tested. Since snapshots can only test HTML and cannot test CSS at all, the "coverage" is only a measure of how many lines of code the useless snapshots have run and not a measure of how many lines of code are _tested_. If we have "100% code coverage" with snapshots, but no visual regression tests, the code coverage measure is a false measurement; what we actually have is a "100% HTML coverage, 0% browser-side JS coverage, and 0% CSS coverage".

Does the "snapshot code coverage" reflect how many components are in storybook? Yes… But snapshots are polluting the test coverage results while they do that. We can easily write a mocha test that checks that all `src/components` have a `*.stories.tsx` file if we really want to capture that data.

**Comment by pmelab** [comment]
> About once every week or two, some non-CSS change in the code base would cause all the auto-generated class names to change even though the CSS didn't change. Because of these frequent false positives, developers got used to just re-generating the snapshots without looking too closely. If there were any real unintended CSS changes, these would have been treated the same as false positives. Tailwind doesn't have auto-generated classes, so this point is moot. In PostCSS, you can work around this problem by configuring webpack to use (during snapshot generation) path-based class names instead of the default hash-based class names that were affected by this problem.

That's what tailwind brings to the table. The class names and CSS properties behind them rarely change. And when they do, then in a very controlled manner (you add /remove new variants via config). The HTML structure becomes a representation of the interface without other side effect vectors.

> Since Jest snapshots only capture the class names, the snapshots don't change even when the class' CSS properties change. Specifically, any change you make in tailwind.config.ts will not change the results of a snapshot. This is super bad. If I change something in the config file mistakenly thinking it will have a small change and it turns out it changes lots of things, the snapshots will never help me discover my mistake. That's exactly what visual regression tests are supposed to do: alert us when we do something stupid.

See above. Yes, tailwind config changes are the one side effect. But its one file, and it would be easy to add a commit-checker that makes sure you don't modify/remove classes that are in use.

>  If we have "100% code coverage" with snapshots, but no visual regression tests, the code coverage measure is a false measurement; what we actually have is a "100% HTML coverage, 0% browser-side JS coverage, and 0% CSS coverage".

To the topic of CSS-coverage, refer to the answers above.

Browser side JS coverage: Inline javascript behaviours are indeed a problem, but at least they show up as "not covered". I touched on this topic in AmazeeLabs/silverback-mono#456. They could be covered with cypress react unit tests, but at the same time, they blur the rendering coverage since they render stuff without actually checking the HTML structure.
If they get refactored into a react hook, they can be covered by unit tests, while the component remains a pure function on the output of the hook. And that's where Storybook and Sinon come in! We can easily stub out hook results to replicate all logic branches of a UI element statically in Storybook and make sure all of these variations stay intact.

My initial idea of this was to combine it with Loki. So Jest can tell us how much of our frontend code is covered, and Loki does the visual integrity check. But it turns out that, if there is no side effect from CSS due to using a static set of classes, Loki does not provide more information than the jest snapshots. And these run 20 times faster.

**Comment by dan2k3k4** [comment]
As John pointed out, we had issues with Jest snapshot tests because the auto-generated classnames kept changing. Sometimes because we had actually changed the CSS inside the CSS Module so it was correct to change the name, but sometimes it changed if the lines were reformatted or similar and this caused confusion and people to ignore or regenerate new snapshots.

I'm happy to try it out again as we may have better luck with it this time.

The snapshots will only work if we own 100% of the DOM (HTML and CSS). If we're pulling in any 3rd party CSS, then we would not be visually testing differences that could happen because the 3rd party CSS styles every 2nd `<div>` element to have margin-top or whatever.

If we could test the exact DOM output per element with parsed CSS then this might bring us closer to true "visual" tests. Such as:

```
  <link
  <style>
    ul:first-child li { padding-top: 1rem; }
  </style>
  <ul className="text-xl">
    <li className="text-blue-500">Active Menu item</li>
    <li className="text-grey-500">Other Menu item</li>
  </ul>
```

Being parsed as:

```
  <ul className="text-xl" style="font-size: 4rem">
    <li className="text-blue-500" style="color: --blue-500; padding-top: 1rem">Active Menu item</li>
    <li className="text-grey-500" style="color: --grey-500">Other Menu item</li>
  </ul>
```

As in the same parsing that the browser ends up with.

**Comment by pmelab** [comment]
3rd party dependencies are a valid point. But nothing stops us from doing some actual visual regression tests for them. We still wouldn't have to render all stories all the time. And also only if the dependency is updated.

**Comment by JohnAlbin** [comment]
I think we are talking about 2 very different purposes for 1 suggested implementation. I disagree 1000% that snapshots can be used as visual regression tests. I can go through a very detailed (read: long) explanation of this, but let's skip this for now.

If you want to implement snapshots as a way to test React components ability to generate the expected HTML (including properties like `class`) and measure its code coverage, I'm all in favor of that. If we can write Storybook stories and have a mocha test auto-generate snapshots, that would be my preferred solution.

Visual regression testing can be a different GitHub issue.
